File: product/ci/addon/artifact/artifact.go
// Copyright 2021 Harness Inc. All rights reserved.
// Use of this source code is governed by the PolyForm Free Trial 1.0.0 license
// that can be found in the licenses directory at the root of this repository, also available at
// https://polyformproject.org/wp-content/uploads/2020/05/PolyForm-Free-Trial-1.0.0.txt.

package artifact

import (
	"encoding/json"
	"fmt"
	"io/ioutil"

	"github.com/pkg/errors"
	pb "github.com/wings-software/portal/product/ci/engine/proto"
)

type (
	// basic structs
	Image struct {
		Image  string `json:"image"`
		Digest string `json:"digest"`
	}
	FileArtifact struct {
		Name string `json:"name"`
		Url  string `json:"url"`
	}

	// artifact data kinds
	DockerV1Data struct {
		RegistryType string  `json:"registryType"`
		RegistryUrl  string  `json:"registryUrl"`
		Images       []Image `json:"images"`
	}
	FileArtifactV1Data struct {
		FileArtifacts []FileArtifact `json:"fileArtifacts"`
	}

	// Artifact formats
	DockerArtifactV1 struct {
		Kind string       `json:"kind"`
		Data DockerV1Data `json:"data"`
	}
	FileArtifactV1 struct {
		Kind string             `json:"kind"`
		Data FileArtifactV1Data `json:"data"`
	}

	Artifact struct {
		*DockerArtifactV1
		*FileArtifactV1
	}
)

func GetArtifactProtoFromFile(artifactFilePath string) (*pb.Artifact, error) {
	var artifactProto = &pb.Artifact{}
	if artifactFilePath != "" {
		artifactObject, err := ParseArtifactFromFile(artifactFilePath)
		if err != nil {
			return nil, err
		} else {
			artifactProto, err = artifactObject.ToArtifactProto()
			if err != nil {
				return nil, err
			}
		}
	}
	return artifactProto, nil
}

func ParseArtifactFromFile(filePath string) (*Artifact, error) {
	json, err := ioutil.ReadFile(filePath)
	if err != nil {
		return nil, errors.Wrap(err, fmt.Sprintf("failed to read artifact file: %s", filePath))
	}
	artifact := &Artifact{}
	err = artifact.UnmarshalJSON(json)
	if err != nil {
		return nil, errors.Wrap(err, fmt.Sprintf("failed to unmarshal artifact json string: %s", json))
	}
	return artifact, nil
}

func (o *Artifact) ToArtifactProto() (*pb.Artifact, error) {
	switch {
	case o.DockerArtifactV1 != nil:
		var imageMetadataList []*pb.DockerImageMetadata
		for _, v := range o.DockerArtifactV1.Data.Images {
			imageMetadataList = append(imageMetadataList, &pb.DockerImageMetadata{
				Image:  v.Image,
				Digest: v.Digest,
			})
		}
		dockerArtifactMetadata := &pb.DockerArtifactMetadata{
			RegistryType: o.DockerArtifactV1.Data.RegistryType,
			RegistryUrl:  o.DockerArtifactV1.Data.RegistryUrl,
			DockerImages: imageMetadataList,
		}
		return &pb.Artifact{
			Metadata: &pb.Artifact_DockerArtifact{dockerArtifactMetadata},
		}, nil
	case o.FileArtifactV1 != nil:
		var fileArtifactList []*pb.FileArtifact
		for _, v := range o.FileArtifactV1.Data.FileArtifacts {
			fileArtifactList = append(fileArtifactList, &pb.FileArtifact{
				Name: v.Name,
				Url:  v.Url,
			})
		}
		fileArtifactMetadata := &pb.FileArtifactMetadata{
			FileArtifacts: fileArtifactList,
		}
		return &pb.Artifact{
			Metadata: &pb.Artifact_FileArtifact{fileArtifactMetadata},
		}, nil
	default:
		return nil, fmt.Errorf("unrecognized kind value %p", o)
	}
}

func (o *Artifact) UnmarshalJSON(data []byte) error {
	var getKind struct{ Kind string }
	if err := json.Unmarshal(data, &getKind); err != nil {
		return err
	}
	switch getKind.Kind {
	case "docker/v1":
		o.DockerArtifactV1 = &DockerArtifactV1{}
		return json.Unmarshal(data, o.DockerArtifactV1)
	case "fileUpload/v1":
		o.FileArtifactV1 = &FileArtifactV1{}
		return json.Unmarshal(data, o.FileArtifactV1)
	default:
		return fmt.Errorf("unrecognized kind value %q", getKind.Kind)
	}

}

func (t Artifact) MarshalJSON() ([]byte, error) {
	switch {
	case t.DockerArtifactV1 != nil:
		return json.Marshal(t.DockerArtifactV1)
	case t.FileArtifactV1 != nil:
		return json.Marshal(t.FileArtifactV1)
	default:
		return json.Marshal(nil)
	}
}

File: 950-delegate-tasks-beans/src/main/java/io/harness/delegate/task/stepstatus/artifact/Artifact.java
/*
 * Copyright 2023 Harness Inc. All rights reserved.
 * Use of this source code is governed by the PolyForm Shield 1.0.0 license
 * that can be found in the licenses directory at the root of this repository, also available at
 * https://polyformproject.org/wp-content/uploads/2020/06/PolyForm-Shield-1.0.0.txt.
 */

package io.harness.delegate.task.stepstatus.artifact;

import com.fasterxml.jackson.annotation.JsonProperty;
import com.fasterxml.jackson.annotation.JsonSubTypes;
import com.fasterxml.jackson.annotation.JsonTypeInfo;
import com.fasterxml.jackson.annotation.JsonValue;
import java.util.List;
import java.util.stream.Collectors;
import lombok.AllArgsConstructor;
import lombok.Data;
import lombok.NoArgsConstructor;
import org.springframework.data.annotation.TypeAlias;

@Data
@NoArgsConstructor
@AllArgsConstructor
class DockerArtifactV1 implements Artifact {
  @JsonProperty("data") private DockerV1Data data;

  @Override
  public Kind getKind() {
    return Kind.DOCKER;
  }

  @Override
  public ArtifactMetadata toArtifactMetadata() {
    return ArtifactMetadata.builder()
        .type(ArtifactMetadataType.DOCKER_ARTIFACT_METADATA)
        .spec(DockerArtifactMetadata.builder()
                  .registryType(data.registryType)
                  .registryUrl(data.registryUrl)
                  .dockerArtifacts(
                      data.images.stream()
                          .map(i -> DockerArtifactDescriptor.builder().imageName(i.image).digest(i.digest).build())
                          .collect(Collectors.toList()))
                  .build())
        .build();
  }

  @Data
  @NoArgsConstructor
  @AllArgsConstructor
  static class DockerV1Data {
    private String registryType;
    private String registryUrl;
    private List<Image> images;

    @Data
    @NoArgsConstructor
    @AllArgsConstructor
    static class Image {
      private String image;
      private String digest;
    }
  }
}

@Data
@NoArgsConstructor
@AllArgsConstructor
class FileArtifactV1 implements Artifact {
  @JsonProperty("data") private FileArtifactV1Data data;

  @Override
  public Kind getKind() {
    return Kind.FILE_UPLOAD;
  }

  @Override
  public ArtifactMetadata toArtifactMetadata() {
    return ArtifactMetadata.builder()
        .type(ArtifactMetadataType.FILE_ARTIFACT_METADATA)
        .spec(
            FileArtifactMetadata.builder()
                .fileArtifactDescriptors(data.fileArtifacts.stream()
                                             .map(f -> FileArtifactDescriptor.builder().url(f.url).name(f.name).build())
                                             .collect(Collectors.toList()))
                .build())
        .build();
  }

  @Data
  @NoArgsConstructor
  @AllArgsConstructor
  static class FileArtifactV1Data {
    private List<FileArtifact> fileArtifacts;

    @Data
    @NoArgsConstructor
    @AllArgsConstructor
    static class FileArtifact {
      private String name;
      private String url;
    }
  }
}

@JsonTypeInfo(use = JsonTypeInfo.Id.NAME, property = "kind")
@JsonSubTypes({
  @JsonSubTypes.Type(value = DockerArtifactV1.class, name = "docker/v1")
  , @JsonSubTypes.Type(value = FileArtifactV1.class, name = "fileUpload/v1")
})
public interface Artifact {
  @TypeAlias("kind")
  enum Kind {
    @JsonProperty("docker/v1") DOCKER("docker/v1"),
    @JsonProperty("fileUpload/v1") FILE_UPLOAD("fileUpload/v1");

    private final String yamlName;

    Kind(String yamlName) {
      this.yamlName = yamlName;
    }

    @JsonValue
    public String getYamlName() {
      return yamlName;
    }
  }

  Kind getKind();
  ArtifactMetadata toArtifactMetadata();
}
